### DISCUSSION TITLE
Is it safe to use `.validated_save()` when duplicating objects?

### DISCUSSION THREAD

**Discussion by rlad78** [nautobot/nautobot#4558] [Q&A] [answered]
Hi all. I'm fairly new when it comes to working with Django, but I've gotten a decent idea of how the Models are structured in Nautobot.

In one of my jobs, I'm looking for a way to duplicate a device multiple times (to create stack members of a virtual chassis). I've seen a lot of recommendations on how to duplicate Model objects, including the [Django recommended way](https://docs.djangoproject.com/en/3.2/topics/db/queries/#copying-model-instances). For example:

```py
import nautobot.dcim.models as dcim_m

class CreateStack(Job):

    ...

    def new_member(self, head_device: dcim_m.Device, ... ) -> dcim_m.Device:
        ....

        head_device_fork = dcim_m.Device.objects.get(pk=head_device.pk)

        head_device_fork.pk = None
        head_device_fork.id = None  # type: ignore
        head_device_fork._state.adding = True
        if head_device.virtual_chassis is not None:
            head_device_fork.vc_position = member_number
            head_device_fork.vc_priority = vc_priority
        head_device_fork.name = base_name + f"-{member_number}"
        head_device_fork.position = rack_position
        head_device_fork.validated_save()

        return head_device_fork
```

I've noticed that, in my dry-runs of this job, new "copied" instances of things like the Interfaces and Power Ports of the original device are created on their own, which is great(?). I didn't expect it to do that all on its own.

The only place it seems to run into trouble is whenever any one-to-one relationships exist, such as a Device having a Primary IPv4 Address assigned. As long as the cloned object has those one-to-one fields cleared out before being saved, it'll go through successfully. However, I haven't handled any Models yet that have a many-to-many relationship. 

While it seems to work well to me, I definitely don't think I have the full picture. Is it safe to copy Model instances this way? Could this cause any database issues down the road?

**Comment by jathanism** [comment, marked as answer]
I would recommend not doing this and instead being more declarative in exactly which fields you would like to carry over in a new instance, for example:

```python
    def new_member(self, head_device: dcim_m.Device, ... ) -> dcim_m.Device:
        ....

         # Explicitly declare the other fields you want from head_device here
        head_device_fork = dcim_m.Device(
            name=head_device.name,
            asset_tag=head_device.asset_tag,
            # etc...
        )
      
        # The rest of this still applies since this is also declarative
        if head_device.virtual_chassis is not None:
            head_device_fork.vc_position = member_number
            head_device_fork.vc_priority = vc_priority
        head_device_fork.name = base_name + f"-{member_number}"
        head_device_fork.position = rack_position
        head_device_fork.validated_save()

        return head_device_fork
```

You might also try a pattern like this that's a little more dynamic but still explicit:

```python
    def new_member(self, head_device: dcim_m.Device, ... ) -> dcim_m.Device
        head_device_fork = dcim_m.Device()
        desired_fields = ["name", "asset_tag", ...] 
        for field_name in desired_fields:
            forked_value = getattr(head_device, field_name)
            setattr(head_device_fork, field_name, forked_value)
        
        # And then the rest of your code.
```

**Reply by rlad78** [reply]
Ah, gotcha. That's what I was planning on doing before. The roadblock I ran into was that I couldn't figure out how to find all the remote fields in order to make sure all the proper values were either copied or replaced. Turns out, you can view them by using `Model._meta.get_fields()`:

```py
        remote_fields = {
            f.name: f.remote_field
            for f in head_device._meta.get_fields()
            if f.is_relation
        }
```

Out of curiosity, I tried to look into how the "Clone" button works for devices. I saw the Models have a `clone_field` class attribute. For example, Device has:
```py
    clone_fields = [
        "device_type",
        "role",
        "tenant",
        "platform",
        "location",
        "rack",
        "status",
        "cluster",
        "secrets_group",
    ]
```

Is there another built-in method that I could possibly use that Nautobot uses as part of its cloning process?

**Reply by glennmatthews** [reply]
So `clone_fields` is intentionally **not** a comprehensive set of fields, it's only the fields that it "makes sense" to duplicate to a new distinct object, leaving the user to fill in the remaining fields with new values as appropriate.

**Reply by rlad78** [reply]
Understood, thanks!
